File: src/devscope/llm/providers/openai.py
from __future__ import annotations

import httpx

from devscope.llm.base import (
    LLMRequest,
    LLMResponse,
    ProviderError,
    RateLimitError,
)
from devscope.secrets import get_secret

# Approximate USD pricing per 1M tokens for common chat models.
# Update as needed; falls back to 0.0 if model not found.
_PRICING_PER_1M: dict[str, tuple[float, float]] = {
    "gpt-4o-mini": (0.15, 0.60),
    "gpt-4o": (2.50, 10.00),
    "gpt-4.1-mini": (0.40, 1.60),
    "gpt-4.1": (2.00, 8.00),
    "o4-mini": (1.10, 4.40),
}


def _cost_usd(model: str, prompt_tokens: int | None, output_tokens: int | None) -> float:
    if model not in _PRICING_PER_1M:
        return 0.0
    in_rate, out_rate = _PRICING_PER_1M[model]
    pt = prompt_tokens or 0
    ot = output_tokens or 0
    return (pt * in_rate + ot * out_rate) / 1_000_000


class OpenAIProvider:
    name = "openai"

    def __init__(
        self,
        *,
        api_key: str | None = None,
        base_url: str = "https://api.openai.com/v1",
        transport: httpx.AsyncBaseTransport | None = None,
        timeout: float = 120.0,
    ) -> None:
        self._api_key = api_key if api_key is not None else (get_secret("OPENAI_API_KEY") or "")
        self._base_url = base_url.rstrip("/")
        self._transport = transport
        self._timeout = timeout

    async def complete(self, req: LLMRequest) -> LLMResponse:
        if not self._api_key:
            raise ProviderError("OPENAI_API_KEY not set")

        payload = {
            "model": req.model,
            "messages": [{"role": "user", "content": req.prompt}],
            "temperature": req.temperature,
            "max_tokens": req.max_tokens,
        }
        headers = {"Authorization": f"Bearer {self._api_key}"}

        async with httpx.AsyncClient(
            base_url=self._base_url,
            transport=self._transport,
            timeout=self._timeout,
        ) as client:
            try:
                resp = await client.post("/chat/completions", json=payload, headers=headers)
            except httpx.HTTPError as exc:
                raise ProviderError(f"openai transport error: {exc}") from exc

        if resp.status_code == 429:
            raise RateLimitError(f"openai 429: {resp.text[:200]}")
        if resp.status_code != 200:
            raise ProviderError(f"openai {resp.status_code}: {resp.text[:200]}")

        data = resp.json()
        usage = data.get("usage", {})
        prompt_tokens = usage.get("prompt_tokens")
        output_tokens = usage.get("completion_tokens")
        return LLMResponse(
            text=data["choices"][0]["message"]["content"],
            prompt_tokens=prompt_tokens,
            output_tokens=output_tokens,
            cost_usd=_cost_usd(req.model, prompt_tokens, output_tokens),
        )

File: src-tauri/src/llm/openai.rs
use async_trait::async_trait;
use serde_json::json;
use std::time::Instant;

use super::{LlmProvider, LlmRequest, LlmResponse};
use crate::error::{AppError, AppResult};

pub struct OpenAIProvider {
    pub api_key: String,
    pub client: reqwest::Client,
}

impl OpenAIProvider {
    pub fn new(api_key: String) -> Self {
        Self {
            api_key,
            client: reqwest::Client::builder()
                .timeout(std::time::Duration::from_secs(120))
                .build()
                .expect("reqwest client"),
        }
    }

    fn price_per_million(model: &str) -> Option<(f64, f64)> {
        match model {
            "gpt-4o-mini" => Some((0.150, 0.600)),
            "gpt-4o" => Some((2.500, 10.000)),
            _ => None,
        }
    }
}

#[async_trait]
impl LlmProvider for OpenAIProvider {
    fn name(&self) -> &'static str {
        "openai"
    }

    async fn call(&self, req: &LlmRequest) -> AppResult<LlmResponse> {
        let started = Instant::now();
        let resp = self
            .client
            .post("https://api.openai.com/v1/chat/completions")
            .bearer_auth(&self.api_key)
            .json(&json!({
                "model": req.model,
                "messages": [{"role": "user", "content": req.prompt}],
            }))
            .send()
            .await
            .map_err(|e| AppError::LlmProvider {
                provider: "openai".into(),
                message: e.to_string(),
            })?;

        if !resp.status().is_success() {
            let status = resp.status();
            let body = resp.text().await.unwrap_or_default();
            return Err(AppError::LlmProvider {
                provider: "openai".into(),
                message: format!("HTTP {status}: {body}"),
            });
        }
        let v: serde_json::Value = resp.json().await?;
        let content = v
            .pointer("/choices/0/message/content")
            .and_then(|x| x.as_str())
            .unwrap_or("")
            .to_string();
        let pt = v.pointer("/usage/prompt_tokens").and_then(|x| x.as_i64());
        let ot = v
            .pointer("/usage/completion_tokens")
            .and_then(|x| x.as_i64());
        let cost = Self::price_per_million(&req.model).map(|(in_p, out_p)| {
            pt.unwrap_or(0) as f64 / 1_000_000.0 * in_p
                + ot.unwrap_or(0) as f64 / 1_000_000.0 * out_p
        });
        if cost.is_none() {
            log::warn!(
                "no pricing entry for openai model {} — cost will not contribute to budget tracking",
                req.model
            );
        }
        Ok(LlmResponse {
            content,
            prompt_tokens: pt,
            output_tokens: ot,
            cost_usd: cost,
            duration_ms: started.elapsed().as_millis() as i64,
            model: req.model.clone(),
            provider: "openai".into(),
        })
    }
}
